### ISSUE TITLE
DeprecationWarning when building with Py_LIMITED_API

### ISSUE THREAD

**Issue by bartoli** [swig/swig#2881]
When running a module generated by swig 4.2, built with Py_LIMITED_API,

The console output gets 3 warnings
`<frozen importlib._bootstrap>:241: DeprecationWarning: builtin type swigvarlink has no __module__ attribute`

For swigPyPacked, SwigPyObject, swigvarlink types.

It seems we could remove them by modifying, in pyrun.swg where the 3 PyType_Spec are filled for thise types:
```
  PyType_Spec spec = {
  #ifndef Py_LIMITED_API
    "SwigPyPacked",
  #else
    SWIG_name ".SwigPyPacked"
  #endif
    sizeof(SwigPyPacked),
```

But in the generated cpp file, SWIG_name is actually defined after those code fragments

I will patch output files on my side to avoid the warning. Do you have an idea on how to make a clean change that could be pushed? Can SWIG_name be defined earlier in the output?

**Comment by ojwb** [comment]
> Can SWIG_name be defined earlier in the output?

```
diff --git a/Source/Modules/python.cxx b/Source/Modules/python.cxx
index 9d195d14..bb832547 100644
--- a/Source/Modules/python.cxx
+++ b/Source/Modules/python.cxx
@@ -782,7 +782,8 @@ public:
     Printf(f_header, "#else\n");
     Printf(f_header, "#  define SWIG_init    init%s\n\n", module);
     Printf(f_header, "#endif\n");
-    Printf(f_header, "#define SWIG_name    \"%s\"\n", module);
+
+    Printf(f_runtime, "#define SWIG_name    \"%s\"\n\n", module);
 
     Printf(f_wrappers, "#ifdef __cplusplus\n");
     Printf(f_wrappers, "extern \"C\" {\n");
```

(BTW, you're missing a comma in your code snippet above.)

**Comment by wsfulton** [comment]
I've done a lot of testing with the limited API and python versions but I haven't seen this. In order to reproduce, could you please provide the value of Py_LIMITED_API that you are using, the version of python you are using to compile against and the version of python you are running with.

**Comment by bartoli** [comment]
This is with python 3.11.7, swig 4.2.0,  Py_LIMITED_API=0x3040000.
It MIGHT also be only when this is a debug build of python and the swig module

The stackoverflow thread where i saw the first solution suggestion: https://stackoverflow.com/questions/66614672/python-limited-api-howto-set-module-attribute-of-extension-type

**Comment by lballabio** [comment]
> I've done a lot of testing with the limited API and python versions but I haven't seen this. 

It only appears when running `python -Walways`, those warnings are hidden by default.

**Comment by bartoli** [comment]
I'll try to propose a PR before monday with my suggested changes and @ojwb's addition

**Comment by bartoli** [comment]
> > I've done a lot of testing with the limited API and python versions but I haven't seen this.
> 
> It only appears when running `python -Walways`, those warnings are hidden by default.

For my case, i think the warning appears because i am using Python debug (python build with --with-pydebug flag)

**Comment by julian-smith-artifex-com** [comment]
I've just seen these warnings when running pytest on PyMuPDF, after adding `#define Py_LIMITED_API 0x03080000` to the underlying MuPDF Python bindings. It's a standard release build, Linux/devuan, Python 3.11.2.

Maybe pytest does the equivalent of `python -Walways`?

**Comment by cdcadman** [comment]
> Maybe pytest does the equivalent of `python -Walways`?

The default behavior in pytest starting with version 3.1 is to catch all warnings and display them at the end (including deprecation warnings, which are by default ignored by python): https://docs.pytest.org/en/stable/how-to/capture-warnings.html

**Comment by julian-smith-artifex-com** [comment]
Thanks @cdcadman, that's good to know.

**Comment by wsfulton** [comment]
These warnings can now be readily seen in the test-suite by setting `PYFLAGS=-Walways`, see e4c13e4.

During the recent improvements to add heap types in #2350, @clintonstimpson's commit 7911683101f9af2bea038ed20fec5ff6e63bcaf0 moved `SWIG_name` to earlier in the generated headers and the module name added to `SwigPyPacked` as suggested here. Alas, the change was mostly reverted in 03a001a5ce4b310145f688214d7342f6fd309203 for the reasons given. I also attempted a similar change for the `swigvarlink` and `SwigPyObject`, but had a few test breakages which I didn't fully understand, except that possibly with the module name being different in each module for the multi-module tests, the types can no longer be found when importing a type from another module. Here are my attempts:
[0002-Fix-python-warning.patch.txt](https://github.com/user-attachments/files/17180022/0002-Fix-python-warning.patch.txt)
[0001-Fix-python-warning.patch.txt](https://github.com/user-attachments/files/17180024/0001-Fix-python-warning.patch.txt)

I'm dropping the swig-4.3 milestone as there is no resolution to date after attempting it and we're about to push out the 4.3 release candidate.

**Comment by ojwb** [comment]
> These warnings can now be readily seen in the test-suite by setting `PYFLAGS=-Walways`

I'm seeing them without any special flags (using Python 3.8.7):

```
checking python testcase member_pointer (with run test)
./member_pointer_runme.py:26: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute
  memberPtr = cvar.perimetervar
./member_pointer_runme.py:29: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute
  check("Square area ", 100.0, do_op(s, cvar.areavar))
./member_pointer_runme.py:30: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute
  check("Square perim", 40.0, do_op(s, cvar.perimetervar))
./member_pointer_runme.py:33: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute
  cvar.areavar = perim_pt
./member_pointer_runme.py:35: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute
  check("Square perimeter", 40.0, do_op(s, cvar.areavar))
./member_pointer_runme.py:39: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute
  memberPtr = AREAPT
checking python testcase member_pointer_const (with run test)
./member_pointer_const_runme.py:26: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute
  memberPtr = cvar.perimetervar
./member_pointer_const_runme.py:29: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute
  check("Square area ", 100.0, do_op(s, cvar.areavar))
./member_pointer_const_runme.py:30: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute
  check("Square perim", 40.0, do_op(s, cvar.perimetervar))
./member_pointer_const_runme.py:33: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute
  cvar.areavar = perim_pt
./member_pointer_const_runme.py:35: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute
  check("Square perimeter", 40.0, do_op(s, cvar.areavar))
./member_pointer_const_runme.py:39: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute
  memberPtr = AREAPT
```

> I'm dropping the swig-4.3 milestone as there is no resolution to date after attempting it and we're about to push out the 4.3 release candidate.

This seems to also trigger for users and there doesn't seem to be a known workaround.  It also doesn't seem to be limited to use of the very latest Python version or to require any special options.  It seems like something we ideally should fix in the next release, hence setting milestone.

**Comment by larsoner** [comment]
FYI one Python-level workaround that might work for people in the meantime (if you're fortunate enough to wrap your SWIG-compiled module in a private namespace!) would be to do something like the following *in the first place you import anything else from your wrapper*:
```
with _warnings.catch_warnings():
    _warnings.simplefilter("ignore")
    from ._openmeeg_wrapper import HeadMat
```
From then on you should be safe to import other stuff from your wrapper as needed I think.

**Comment by fergusondave** [comment]
In 3.10.7 Debug mode on Windows, I'm seeing this issue with swig-4.3 even though **I'm not defining `Py_LIMITED_API`**.

According to this [note](https://github.com/python/cpython/issues/124182#issuecomment-2360945364), the issue is the format of the `tp_name` field in `PyTypeObject`.  It should be `"myModule.MyClass"` for static types (which are used for -builtin).  Even the Python 2 docs for [`tp_name`](https://docs.python.org/2/c-api/typeobj.html#c.PyTypeObject.tp_name) specify this format.

I'm not using -builtin, so `SWIG_HEAPTYPES` gets defined, thus swig uses [dynamic types](https://docs.python.org/3/c-api/typeobj.html#heap-types).  The warning is generated by the call to `PyType_FromSpec` (which eventually calls `PyType_FromModuleAndSpec`), and is in newer versions of Python as well.  It looks like the check was introduced in 3.5.0.

For dynamic types, this call will setup `__module__` in the dict as the `tp_name` docs specify, but it looks like `tp_name` will still have the module dot part.  So, if swig wants to setup _just_ the type name in `tp_name`, I think it would need to create the dynamic type's dict and set `__module__` before calling `PyType_FromSpec`.  Something something slots, might be an issue with `Py_LIMITED_API`, etc.

TLDR, if I prepend `SWIG_name "."` to every occurrence of "SwigPyPacked", "swigvarlink", and "SwigPyObject" (e.g. `SWIG_name ".SwigPyPacked"`) in my generated wrapper, the warnings go away.

I don't know if changing this would cause conflicts when you have multiple swig-generated modules or backwards-compatibility issues with the same.

**Comment by jschueller** [comment]
we cannot use `SWIG_name ".SwigPyPacked"` because SWIG_name is not known with the external_runtime use-case.
Seems just ".SwigPyPacked" works, see #3160

**Comment by jschueller** [comment]
fair enough
did you recall what was failing ? there was a gotcha for stable api but now every test pass

**Comment by jschueller** [comment]
@wsfulton my patch (#3160, all tests ok)  is indeed similar to yours (0002-Fix-python-warning.patch.txt)  
except for the PyUnicode_CompareWithASCIIString line which I kept unchanged unlike you, maybe that why yours failed

**Comment by wsfulton** [comment]
Yes, my patch I had to revert had SWIG_name in it: 03a001a5ce4b310145f688214d7342f6fd309203. I'm not sure what a simple `.` prefix means, is it a real workaround / fix to requiring a module name. It isn't really providing a module name!

**Comment by wsfulton** [comment]
`.SwigPyObject` is a really odd class name. It does provide an empty `__module__` name. Just doesn't feel right at all to me, the module name should be set. This is the 'class' example in Examples/python/class, using patch in #3160 and `-builtin`:

```python
>>> import example
>>> example.Square.__bases__
(<class 'example.Shape'>,)
>>> example.Square.__bases__[0]
<class 'example.Shape'>
>>> example.Square.__bases__[0].__bases__
(<class '.SwigPyObject'>,)
>>> example.Square.__bases__[0].__bases__[0]
<class '.SwigPyObject'>
>>> example.Square.__bases__[0].__bases__[0].__bases__
(<class 'object'>,)
>>> example.Square.__bases__[0].__bases__[0].__module__
''
>>> .SwigPyObject.__class__
  File "<python-input-8>", line 1
    .SwigPyObject.__class__
    ^
SyntaxError: invalid syntax
```
In comparison, current master version is:
```python
>>> example.Square.__bases__[0]
<class 'example.Shape'>
>>> example.Square.__bases__[0].__bases__
(<class 'SwigPyObject'>,)
>>> example.Square.__bases__[0].__bases__[0]
<class 'SwigPyObject'>
>>> example.Square.__bases__[0].__bases__[0].__bases__
(<class 'object'>,)
>>> example.Square.__bases__[0].__bases__[0].__module__
'builtins'
```
So SwigPyObject is in the 'builtins' module, like list, tuple, etc!

**Comment by jschueller** [comment]
I can try with the builtins module to see what happens

**Comment by wsfulton** [comment]
Modifying the builtins as provided by Python could be interpreted as something between bad practice and anarchic, so we should fix that.

I've been trying to figure out what to do here as I think the warning is trying to encourage types to be in a module. All the wrapped classes go into the user specified module, so putting them in this module could be the obvious thing to do. However, these 3 types are need to be shared amongst different modules when multiple modules are used with `%import`, so some sort of other solution is needed in the multiple module scenario.

I've decided to put these 3 types into the runtime module that already exists. The module is currently called `swig_runtime_data5`, where 5 is the version of the runtime as defined by `SWIG_RUNTIME_VERSION`. The module contains the Capsule which holds the C pointer for the runtime data structures shared across multiple modules.

**Comment by wsfulton** [comment]
Fixed now including some documentation on these 3 types being added to the runtime module. See commits 766b7d45ac2cc151919a30a9844cb079b5b98bd9...d966fd3ba18370b5bf34b67ff376d88fb6d3e65b
